src/python/pants/java/nailgun_io.py

```diff
@@ -205,6 +209,73 @@ def __init__(self, in_fds, sock, chunk_types, chunk_eof_type, buf_size=None, sel
   def _assert_aligned(self, *iterables):
     assert len({len(i) for i in iterables}) == 1, 'inputs are not aligned'
 
+  def _handle_closed_input_stream(self, fileno):
+    # We've reached EOF.
+    try:
+      if self._chunk_eof_type is not None:
+        NailgunProtocol.write_chunk(self._socket, self._chunk_eof_type)
+    finally:
+      try:
+        os.close(fileno)
+      finally:
+        self.stop_reading_from_fd(fileno)
+
+  def stop_reading_from_fd(self, fileno):
+    self._in_fds.remove(fileno)
+
+  def run(self):
+    while self._in_fds and not self.is_stopped:
+      readable, _, errored = select.select(self._in_fds, [], self._in_fds, self._select_timeout)
+      self.do_run(readable, errored)
+
+  def do_run(self, readable, errored):
+    """Represents one iteration of the infinite reading cycle. Subclasses should override this."""
+    if readable:
+      for fileno in readable:
+        data = os.read(fileno, self._buf_size)
+        if not data:
+          self._handle_closed_input_stream(fileno)
+        else:
+          NailgunProtocol.write_chunk(
+            self._socket,
+            self._fileno_chunk_type_map[fileno],
+            data
+          )
+
+    if errored:
+      for fileno in errored:
+        self.stop_reading_from_fd(fileno)
+
+
+class PipedNailgunStreamWriter(NailgunStreamWriter):
+  """
+  Represents a NailgunStreamWriter that reads from a pipe.
+  """
+
+  def __init__(self, pipes, socket, chunk_type, *args, **kwargs):
+    self._pipes = pipes
+    in_fds = tuple(pipe.read_fd for pipe in pipes)
+    super(PipedNailgunStreamWriter, self).__init__(in_fds, socket, chunk_type, *args, **kwargs)
+
+  def do_run(self, readables, errored):
+    """
+    Overrides the superclass.
+
+    Wraps the running logic of the parent class to handle pipes that have been closed on the write end.
+    If no file descriptors are readable (i.e. there is no more to read from any pipe for now),
+    it will check each of its pipes. If a pipe is not writable, it will interpret that the writer class
+    does not want to write any more, and so it will remove that pipe from the available pipes to read from.
+    When there are no more pipes to read from, it will stop.
+    """
+    super(PipedNailgunStreamWriter, self).do_run(readables, errored)
+    if not readables:
```

Comment:
Should this filter `readables` based on what the superclass did at all?

Reply:
Done :) We want the pipe info to supersede the superclass, so I just moved the super call to the bottom.